src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -196,6 +213,14 @@ double LMNNFunction<MetricType>::Evaluate(const arma::mat& transformation,
 {
   double cost = 0;
 
+  // Ensure cache transformation cube has correct size.
+  if (transformationOldPoint.n_elem == 0)
+  {
+    transformationOldPoint.set_size(transformation.n_rows,
+        transformation.n_cols, dataset.n_cols);
+    transformationOldPoint.fill(arma::datum::nan);
+  }
```

Comment:
I think this means your approach is that you will store one transformation matrix per point?  That will be very large in memory.  I'd suggest instead a `std::vector<arma::mat>` of old transformation matrices, and then a `arma::Row<size_t>` to track the index that is used for each point.  (I know that earlier I suggested changing from a `std::vector<arma::mat>` to an `arma::cube`, but that was for a case where the number of slices in the cube wouldn't change.  Here, I think it would be better for memory to only store as many transformation matrices as we need.  Also, I guess each call to Shuffle() we can check for unused transformation matrices and either remove them or just call `clear()`.)

Reply:
I have some doubts regarding this -
This is mainly done for batch support, the reason why I have done this because of the changing batch size, Let''s suppose in next iteration batch size changed, then I am not able to see how will we keep  track of the transformationOld now. Please explain a little bit more, maybe I am missing something here.

Reply:
Right, so the idea is this.  Hold an `arma::Row<size_t>` called `lastTransformationIndices` (or something), and hold a `std::vector<arma::mat>` called `oldTransformationMatrices` (or something).  Then, when we do a batch:

```
oldTransformationMatrices.push_back(transformationMatrix);
lastTransformationIndices.subvec(begin, begin + batchSize - 1) = oldTransformationMatrices.size();
```

Then, when you need to get the transformation matrix for a point `i`, do `oldTransformationMatrices[lastTransformationIndices[i]]`.  Do you see what I mean?  There will be a little bit of extra complexity here but that is the basic idea.

Reply:
The update process is nasty and will require some extra computation time at each Evaluate() call
```cpp
  // Update & remove unused cache matrices.
  oldTransformationMatrices.push_back(transformation);
  lastTransformationIndices.cols(begin, begin + batchSize - 1) =
      oldTransformationMatrices.size() - 1;

  size_t maxIndex = lastTransformationIndices.max();
  arma::uvec matricesUsed = arma::unique(lastTransformationIndices);
  for (size_t i = 0, j = 0, l = 0; i <= maxIndex; i++)
  {
    if (i != matricesUsed(j))
    {
      oldTransformationMatrices.erase(i - l);
      l++;
    }
    else
      j++;
  }
  // Normalize lastTransformationIndices.
  Maybe NormalizeLabels() can be used for this.
```


So, Do you think it is worth the tradeoff for space? And I think we are using the same amount of space for `evalOld` as well.

Reply:
Right, let's consider using a slightly different strategy, holding the following auxiliary variables:

```
// It might be useful to come up with shorter names, these are really long...

// This will hold each of the transformation matrices.
std::vector<arma::mat> oldTransformationMatrices;
// This will hold how many points are using each transformation matrix.
std::vector<size_t> oldTransformationCounts;
// This will hold which transformation matrix each point is using.
arma::Row<size_t> lastTransformationIndices;
```

Then, whenever we call `Evaluate(..., begin, batchSize)`, we can do the following:

```
// Are there any empty transformation matrices?
size_t index = oldTransformationMatrices.size();
for (size_t i = 0; i < oldTransformationCounts.size(); ++i)
{
  if (oldTransformationCounts[i] == 0)
  {
    index = i; // Reuse this index.
    break;
  }
}

// Did we find an unused matrix?  If not, we have to allocate new space.
if (index == oldTransformationMatrices.size())
{
  oldTransformationMatrices.push_back(transformationMatrix);
  oldTransformationCounts.push_back(0);
}
else
{
  oldTransformationMatrices[index] = transformationMatrix;
}
    
// Update all the transformation indices.
for (size_t i = begin; i < begin + batchSize; ++i)
  --oldTransformationCounts[lastTransformationIndices[i]];
lastTransformationIndices.cols(begin, begin + batchSize - 1) = index;
oldTransformationCounts[index] += batchSize;
```

During initialization for the whole procedure I'd suggest setting `oldTransformationMatrices` to have one empty matrix and set all the `lastTransformationIndices` to `0` and the counts accordingly.

And, when shuffling is done, we just need to modify `lastTransformationIndices`, which will be very similar to the existing shuffling code.

This is an improvement over the idea above, because we don't need to call `arma::unique()` every call, which is very expensive.  Instead the cost of this approach, at worst, is one transformation matrix copy followed by `2 * batchSize + 1` assignments, which is pretty trivial compared to the rest of the computation.

Reply:
Thanks!  That's a really good algorithm to handle the situation :)